### PR TITLE
formatter: add a formatter that returns a google::protobuf::Struct rather than a string

### PR DESCRIPTION
**Pull request by itamarkam**
Commit Message: add a formatter that returns google::protobuf::Struct rather than a string
Additional Description: the logic was already implemented in the JSON (string) formatter, which created a formatted struct and converted it into a string. The JSON formatter was split into the aforementioned Struct formatter, and a wrapper that converts the struct into a string. Note that the new formatter is still not used anywhere (will be in the [OTLP logger](https://github.com/envoyproxy/envoy/issues/13801))
Risk Level: Low (refactoring)
Testing: Unit tests (JSON formatter tests were converted to Struct formatter tests and a simpler test was added to the JSON formatter).
Docs Changes: None
Release Notes: None
Platform Specific Features: None
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]

@omriz @yanavlasov @htuch @edrukh

### CODE REVIEW

**Comment by itamarkam** [comment]
@yanavlasov this is a small refactoring, to be used later by the OT logger - I wanted to get the hang of Envoy repo PRs with something relatively simple. If you prefer that I wait with this change and send it together with the actual logger implementation, I'll do that.

File: test/common/formatter/substitution_formatter_speed_test.cc [hunk 1]
```diff
@@ -63,6 +81,47 @@ static void BM_AccessLogFormatter(benchmark::State& state) {
...
+static void BM_StructAccessLogFormatter(benchmark::State& state) {
+  std::unique_ptr<Envoy::TestStreamInfo> stream_info = makeStreamInfo();
+  std::unique_ptr<Envoy::Formatter::StructFormatter> struct_formatter = makeStructFormatter(false);
+
+  size_t output_bytes = 0;
+  Http::TestRequestHeaderMapImpl request_headers;
+  Http::TestResponseHeaderMapImpl response_headers;
+  Http::TestResponseTrailerMapImpl response_trailers;
+  std::string body;
+  for (auto _ : state) {
```
Comment (itamarkam) [inline test/common/formatter/substitution_formatter_speed_test.cc:94]: This (and line 115) triggers a clang-tidy error, but I just copied it from other tests in this file.

Comment (zuercher) [reply test/common/formatter/substitution_formatter_speed_test.cc:94]: `for (auto _ : state) { // NOLINT` is how this has been solved elsewhere. Clang-tidy is only applied to changes in our CI, which is why others in this file didn't trigger it. You may have to fix those as well.

**Review by zuercher — COMMENTED** [review]
Thanks! This is just a quick initial pass.

File: source/common/formatter/substitution_formatter.cc [hunk 2]
```diff
@@ -154,18 +154,18 @@ JsonFormatterImpl::toFormatMap(const ProtobufWkt::Struct& json_format) const {
       output->emplace(pair.first, toFormatMap(pair.second.struct_value()));
       break;
     default:
-      throw EnvoyException(
-          "Only string values or nested structs are supported in the JSON access log format.");
+      throw EnvoyException("Only string values or nested structs are supported in the Struct/JSON "
+                           "access log format.");
```
Comment (zuercher) [inline source/common/formatter/substitution_formatter.cc:158]: Consider changing this to "... in structured access log formats".

File: source/common/formatter/substitution_formatter.h [hunk 3]
```diff
@@ -107,12 +107,46 @@ class FormatterImpl : public Formatter {
...
+  // Although not required for Struct/JSON, it is nice to have the order of
+  // properties preserved between the format and the log entry, thus std::map.
+  using StructFormatMap = std::map<std::string, StructFormatMapValue>;
+  using StructFormatMapPtr = std::unique_ptr<StructFormatMap>;
+  struct StructFormatMapWrapper {
+    StructFormatMapPtr value_;
+  };
+
+  bool omit_empty_values_;
+  bool preserve_types_;
```
Comment (zuercher) [inline source/common/formatter/substitution_formatter.h:140]: I think these bools be const as well.

File: source/common/formatter/substitution_formatter.h [hunk 4]
```diff
@@ -107,12 +107,46 @@ class FormatterImpl : public Formatter {
...
+  const StructFormatMapWrapper struct_output_format_;
+  StructFormatMapWrapper toFormatMap(const ProtobufWkt::Struct& struct_format) const;
```
Comment (zuercher) [inline source/common/formatter/substitution_formatter.h:142]: style: methods after types, before member variables.

File: source/common/formatter/substitution_formatter.h [hunk 5]
```diff
@@ -122,27 +156,7 @@ class JsonFormatterImpl : public Formatter {
...
-  bool preserve_types_;
-  const JsonFormatMapWrapper json_output_format_;
-
-  ProtobufWkt::Struct toStruct(const Http::RequestHeaderMap& request_headers,
-                               const Http::ResponseHeaderMap& response_headers,
-                               const Http::ResponseTrailerMap& response_trailers,
-                               const StreamInfo::StreamInfo& stream_info,
-                               absl::string_view local_reply_body) const;
-  JsonFormatMapWrapper toFormatMap(const ProtobufWkt::Struct& json_format) const;
+  StructFormatter struct_formatter_;
```
Comment (zuercher) [inline source/common/formatter/substitution_formatter.h:159]: const?

**Review by zuercher — COMMENTED** [review]
Thanks! Some small comments, but otherwise looks good.

**Comment by mattklein123** [comment]
I think this was tagged for API remove accidentally so I'm removing myself.

**Comment by yanavlasov** [comment]
LGTM module @zuercher comments.

**Review by yanavlasov — APPROVED** [review]

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**